cocotb/__init__.py

```diff
@@ -269,18 +269,40 @@ def _sim_event(level, message):
     from cocotb.result import TestFailure, SimFailure
 
     if level is SIM_TEST_FAIL:
-        scheduler.log.error("Failing test at simulator request")
-        scheduler.finish_test(TestFailure("Failure from external source: %s" %
-                              message))
+        if regression_manager is None:  # pragma: no cover
+            # Test failure before regression manager started. Should never happen.
+            cocotb.log.error("Test failed before Regression Manager started: " + message)
+
+        elif not regression_manager.finished:
+            # Test failure during a test, fail it
+            msg = "Test failure from external source: " + message
+            cocotb.log.error(msg)
+            scheduler.finish_test(TestFailure(msg))
+
+        else:  # pragma: no cover
+            # Test failure after the regression manager finishes. Should never happen.
+            msg = "Received test failure, but all tests have finished: " + message
+            cocotb.log.info(msg)
+
     elif level is SIM_FAIL:
-        # We simply return here as the simulator will exit
-        # so no cleanup is needed
-        msg = ("Failing test at simulator request before test run completion: "
-               "%s" % message)
+        if regression_manager is None:  # pragma: no cover
+            # Sim failure/end before regression manager started. Should never happen.
+            cocotb.log.error("Regression ended before Regression Manager started: " + message)
+
```

Comment:
This will happen any time `embed_sim_init` fails or `_initialise_testbench` raises an Exception before `regression_manager` is initialized. An easy way to test this is to set `TOPLEVEL` to a bad value.

Reply:
Icarus will fail to find the top level and error before cocotb is loaded.

Reply:
Added a test for this in the most recent commit.

Reply:
`TOPLEVEL` discovery is in `_initialise_testbench`:

```
     -.--ns INFO     cocotb.gpi                         ..mbed/gpi_embed.cpp:97   in set_program_name_in_venv        Using Python virtual environment interpreter at .../env/bin/python
     -.--ns INFO     cocotb.gpi                         ../gpi/GpiCommon.cpp:106  in gpi_print_registered_impl       VPI registered
     -.--ns INFO     cocotb.gpi                         ..mbed/gpi_embed.cpp:244  in embed_sim_init                  Python interpreter initialized and cocotb loaded!
     0.00ns INFO     cocotb                                      __init__.py:212  in _initialise_testbench           Running on Icarus Verilog version 12.0 (devel)
     0.00ns INFO     cocotb                                      __init__.py:218  in _initialise_testbench           Running tests with cocotb v1.5.0.dev0 from .../env/lib/python3.8/site-packages/cocotb
     0.00ns INFO     cocotb                                      __init__.py:239  in _initialise_testbench           Seeding Python random module with 1603472827
     0.00ns ERROR    cocotb.gpi                                  VpiImpl.cpp:495  in get_root_handle                 VPI: Couldn't find root handle garbage
     0.00ns ERROR    cocotb.gpi                                  VpiImpl.cpp:501  in get_root_handle                 VPI: Toplevel instances: garbage != $unit...
     0.00ns ERROR    cocotb.gpi                                  VpiImpl.cpp:501  in get_root_handle                 VPI: Toplevel instances: garbage != dff...
     0.00ns ERROR    cocotb.gpi                                GpiCommon.cpp:288  in gpi_get_root_handle             No root handle found
Traceback (most recent call last):
  File ".../env/lib/python3.8/site-packages/cocotb/__init__.py", line 250, in _initialise_testbench
    raise RuntimeError("Can not find root handle ({})".format(root_name))
RuntimeError: Can not find root handle (garbage)
     0.00ns ERROR    cocotb.gpi                                gpi_embed.cpp:314  in embed_sim_init                  cocotb initialization failed - exiting
     0.00ns ERROR    cocotb                                      __init__.py:290  in _sim_event                      Regression ended before Regression Manager started: Failed to initialize GPI user.
ERROR: results.xml was not written by the simulation!
```

Reply:
Clean that directory and run it again and you'll see what I mean.

Reply:
Remove the pragma and update the comment?